numba/core/entrypoints.py

```diff
@@ -1,31 +1,62 @@
 import logging
+import re
 import warnings
 
-from pkg_resources import iter_entry_points
+from numba.core.config import PYVERSION
+
+if PYVERSION < (3, 8):
+    try:
+        import importlib_metadata
+    except ImportError as ex:
+        raise ImportError(
+            "importlib_metadata backport is required for Python version < 3.8, "
+            "try:\n"
+            "$ conda/pip install importlib_metadata"
+        ) from ex
+else:
+    from importlib import metadata as importlib_metadata
+
 
 _already_initialized = False
 logger = logging.getLogger(__name__)
 
 
 def init_all():
-    '''Execute all `numba_extensions` entry points with the name `init`
+    """Execute all `numba_extensions` entry points with the name `init`
 
     If extensions have already been initialized, this function does nothing.
-    '''
+    """
     global _already_initialized
     if _already_initialized:
         return
 
     # Must put this here to avoid extensions re-triggering initialization
     _already_initialized = True
 
-    for entry_point in iter_entry_points('numba_extensions', 'init'):
-        logger.debug('Loading extension: %s', entry_point)
-        try:
-            func = entry_point.load()
-            func()
-        except Exception as e:
-            msg = "Numba extension module '{}' failed to load due to '{}({})'."
-            warnings.warn(msg.format(entry_point.module_name, type(e).__name__,
-                                     str(e)), stacklevel=2)
-            logger.debug('Extension loading failed for: %s', entry_point)
+    for entry_point in importlib_metadata.entry_points().get(
+        "numba_extensions", ()
+    ):
+        if entry_point.name == "init":
+            logger.debug('Loading extension: %s', entry_point)
+            try:
+                func = entry_point.load()
+                func()
+            except Exception as e:
+                # entry_point.module was only added in Python 3.9. The
+                # backported importlib_metadata (used on Python 3.7) also adds
+                # it. So, on Python 3.8 we need to duplicate the logic used to
+                # extract the module name.
+                if PYVERSION == (3, 8):
```

Comment:
Why not just use `importlib_metadata` on Python 3.8 too?

Reply:
Good point - done.